### PR TITLE
[lexical-react] Bug Fix: LexicalTypeaheadMenuPlugin SSR error: ReferenceError: document is not defined

### LINKED ISSUE #6716: Bug: LexicalTypeaheadMenuPlugin  SSR error: `ReferenceError: document is not defined`
It was mentioned [before](https://github.com/facebook/lexical/issues/1177#issuecomment-1022094859)  that you have a util to check if the DOM can be used. I don't think this check is being used in the `LexicalTypeaheadMenuPlugin`, considering I'm getting 
this error:

```
ReferenceError: document is not defined
    at useMenuAnchorRef (....../node_modules/@lexical/react/LexicalTypeaheadMenuPlugin.dev.mjs:337:74)
    at LexicalTypeaheadMenuPlugin (....../node_modules/@lexical/react/LexicalTypeaheadMenuPlugin.dev.mjs:534:28)
```

I'm running a remix project and trying to import the playground. I ran into all sorts of problems, so I used `React.lazy` to import the component only in the frontend. That solved all of the issues, however the above error is logged during server rendering.

### PR DESCRIPTION
**Pull request by bedre7**
## Description

Used `CAN_USE_DOM` to check if the DOM can be used when accessing the document object in the `useMenuAnchorRef` hook. This prevents runtime errors when the hook is used by a plugin in SSR.

Closes #6716 
## Test plan
Unit test has been added for `useMenuAnchorRef`.

### CODE REVIEW

File: packages/lexical-react/src/LexicalTypeaheadMenuPlugin.tsx [hunk 1]
```diff
@@ -295,12 +295,14 @@ export function LexicalTypeaheadMenuPlugin<TOption extends MenuOption>({
...
-  return resolution === null || editor === null ? null : (
+  return resolution === null ||
+    editor === null ||
+    anchorElementRef.current === null ? null : (
     <LexicalMenu
       close={closeTypeahead}
       resolution={resolution}
       editor={editor}
-      anchorElementRef={anchorElementRef}
+      anchorElementRef={anchorElementRef as React.MutableRefObject<HTMLElement>}
```
Comment (etrepum) [inline packages/lexical-react/src/LexicalTypeaheadMenuPlugin.tsx:305]:
```suggestion
      anchorElementRef={anchorElementRef}
```

File: packages/lexical-react/src/LexicalNodeMenuPlugin.tsx [hunk 2]
```diff
@@ -110,12 +110,14 @@ export function LexicalNodeMenuPlugin<TOption extends MenuOption>({
...
-  return resolution === null || editor === null ? null : (
+  return anchorElementRef.current === null ||
+    resolution === null ||
+    editor === null ? null : (
     <LexicalMenu
       close={closeNodeMenu}
       resolution={resolution}
       editor={editor}
-      anchorElementRef={anchorElementRef}
+      anchorElementRef={anchorElementRef as React.MutableRefObject<HTMLElement>}
```
Comment (etrepum) [inline packages/lexical-react/src/LexicalNodeMenuPlugin.tsx:120]:
```suggestion
      anchorElementRef={anchorElementRef}
```

File: packages/lexical-react/src/LexicalContextMenuPlugin.tsx [hunk 3]
```diff
@@ -144,12 +144,14 @@ export function LexicalContextMenuPlugin<TOption extends MenuOption>({
...
-  return resolution === null || editor === null ? null : (
+  return anchorElementRef.current === null ||
+    resolution === null ||
+    editor === null ? null : (
     <LexicalMenu
       close={closeNodeMenu}
       resolution={resolution}
       editor={editor}
-      anchorElementRef={anchorElementRef}
+      anchorElementRef={anchorElementRef as MutableRefObject<HTMLElement>}
```
Comment (etrepum) [inline packages/lexical-react/src/LexicalContextMenuPlugin.tsx:154]:
```suggestion
      anchorElementRef={anchorElementRef}
```

**Review by etrepum — COMMENTED** [review]
I think that the LexicalMenu anchorElementRef type should change to match the new expectations rather than using a cast. It's a mutable object, the contents can change from not-null to null at any time (at least per the type contract)

File: packages/lexical-react/src/shared/LexicalMenu.ts [hunk 4]
```diff
@@ -531,7 +537,7 @@ export function useMenuAnchorRef(
         }
       }
 
-      if (!containerDiv.isConnected) {
+      if (!containerDiv.isConnected && parent !== undefined) {
```
[the remark this answers is not in the document]
Comment (bedre7) [reply packages/lexical-react/src/shared/LexicalMenu.ts]: Done

File: packages/lexical-react/src/shared/LexicalMenu.ts [hunk 5]
```diff
@@ -481,12 +482,17 @@ export function useMenuAnchorRef(
...
   shouldIncludePageYOffset__EXPERIMENTAL: boolean = true,
-): MutableRefObject<HTMLElement> {
+): MutableRefObject<HTMLElement | null> {
   const [editor] = useLexicalComposerContext();
-  const anchorElementRef = useRef<HTMLElement>(document.createElement('div'));
+  const anchorElementRef = useRef<HTMLElement | null>(
+    CAN_USE_DOM ? document.createElement('div') : null,
+  );
   const positionMenu = useCallback(() => {
+    if (anchorElementRef.current === null) {
```
[the remark this answers is not in the document]
Comment (bedre7) [reply packages/lexical-react/src/shared/LexicalMenu.ts]: Done

File: packages/lexical-react/src/LexicalTypeaheadMenuPlugin.tsx (hunk 1 above)
Comment (bedre7) [reply packages/lexical-react/src/LexicalTypeaheadMenuPlugin.tsx]: Done

File: packages/lexical-react/src/LexicalNodeMenuPlugin.tsx (hunk 2 above)
Comment (bedre7) [reply packages/lexical-react/src/LexicalNodeMenuPlugin.tsx]: Done

File: packages/lexical-react/src/LexicalContextMenuPlugin.tsx (hunk 3 above)
Comment (bedre7) [reply packages/lexical-react/src/LexicalContextMenuPlugin.tsx]: Done

File: packages/lexical-react/src/shared/LexicalMenu.ts [hunk 6]
```diff
@@ -267,7 +268,7 @@ export function LexicalMenu<TOption extends MenuOption>({
 }: {
   close: () => void;
   editor: LexicalEditor;
-  anchorElementRef: MutableRefObject<HTMLElement>;
+  anchorElementRef: React.MutableRefObject<HTMLElement | null>;
```
Comment (etrepum) [inline packages/lexical-react/src/shared/LexicalMenu.ts:271]:
```suggestion
  anchorElementRef: MutableRefObject<HTMLElement | null>;
```

**Review by etrepum — DISMISSED** [review]
I think this is good to go, I will apply the small suggested nit and assuming the e2e test suite looks good then it will be merged. Thank you!

**Review by etrepum — APPROVED** [review]

**Pull request merged**